File: src/ModalTableaux.hs
module Tableaux where

import Data.Set (Set)
import qualified Data.Set as Set
import Data.Tree (Tree)
import qualified Data.Tree as Tree
import Control.Monad (guard)

import Prop

-- | A tableaux is a tree with nodes
-- | sets of propositions, their respective world id, and sat (i.e. ~closed)
type NodeType = (Set Phi, Integer, Bool)
type Tableaux = Tree NodeType

data SetStatus = NonAtomic | Open | Closed deriving (Eq, Show)

tableaux :: Phi -> Tableaux
tableaux sp = (build . Set.singleton) sp

{-
start with initial world 0
exhaust type 1 rules (simple ST + put ¬ inside for box and diamond)
construct new world (prev + 1) for each diamond p => p is true in this world
box q will be true in all succs => copied without box
also, because we are in K (p -> ☐p), all literals in a world are copied in the succs of that world
-}
build :: Set Phi -> Tableaux
build sp = case describeSet sp of
    Open      -> Tree.Node (sp, False) []
    Closed    -> Tree.Node (sp, True)  []
    NonAtomic -> Tree.Node (sp, wId, cls) rest
    where
        (nonAtoms, _) = Set.partition (not . isAtom) sp
        m    = Set.findMin nonAtoms
        sp'  = Set.delete m sp
        rest = map (build . Set.union sp') (branchOn m)
        cls  = and $ map isNodeClosed rest

-- | Check if formulas are atomic
isAtom :: Phi -> Bool
isAtom (Var _)       = True
isAtom (Not (Var _)) = True
isAtom _             = False

allAtoms :: Set Phi -> Bool
allAtoms sp = Set.null $ Set.filter (not . isAtom) sp


-- | Check if there are complements in a set of formulas
complement :: Phi -> Phi
complement (Not p) = p
complement p       = Not p

existsComplements :: Set Phi -> Bool
existsComplements sp
    | Set.size sp <= 1 = False
    | otherwise        = if Set.member (complement m) sp' then True else existsComplements sp'
    where
        m   = Set.findMin sp
        sp' = Set.deleteMin sp


-- | Get the closed-state of a node
isNodeClosed :: Tableaux -> Bool
isNodeClosed (Tree.Node (_, _, c) _) = c


-- | Describe a set based on contained formulas
-- | Closed     <-> ∃ p, ~p in the set
-- | Open       <-> non-closed, atomic elements => sat
-- | NonAtomic  <-> ∃ f that can be further expanded
describeSet :: Set Phi -> SetStatus
describeSet sp = if existsComplements sp
    then Closed
    else if (not . allAtoms) sp then NonAtomic else Open


-- | Construction rules

-- | Construct separate subsets
beta :: [Phi] -> [Set Phi]
beta xs = map Set.singleton xs

-- | Construct a single set
alpha :: [Phi] -> [Set Phi]
alpha xs = [Set.fromList xs]

-- | Node expansion rules
branchOn :: Phi -> [Set Phi]

-- | No ramification, just _expand_ the formula
branchOn (And p q)       = alpha [p, q]
branchOn (Not (Or p q))  = alpha [Not p, Not q]
branchOn (Not (Imp p q)) = alpha [p, Not q]
branchOn (Iff p q)       = alpha [Imp p q, Imp q p]
branchOn (Not (Not p))   = alpha [p]

-- | Ramification
branchOn (Or p q)        = beta [p, q]
branchOn (Imp p q)       = beta [Not p, q]
branchOn (Not (And p q)) = beta [Not p, Not q]
branchOn (Not (Iff p q)) = beta [Not (Imp p q), Not (Imp q p)]

-- | Modal logic
branchOn (Not (Nec p))   = alpha [Pos (Not p)]
branchOn (Not (Pos p))   = alpha [Nec (Not p)]


-- | Tree printing utils
closedStr True  = " {X} "
closedStr False = " {O} "

showNode :: NodeType -> String
showNode (sp, cls) = show (Set.toList sp) ++ (closedStr cls)

-- | nodes -> string nodes with showNode, then show all tree
showTableaux :: Tableaux -> String
showTableaux t = (Tree.drawTree . fmap showNode) t

File: src/ModalTableaux.py
from enum import Enum

from copy import copy
from functools import reduce

from Prop import *
from Utils import partition, findMin, getName



class SetStatus(Enum):
    NonAtomic = 1
    Open      = 2
    Closed    = 3


class Tree:
    def __init__(self, sp, closed, kids, world=0):
        self.sp     = sp
        self.closed = closed
        self.kids   = kids
        self.world  = world

    def __repr__(self):
        return "<semantic tableaux>"

    def __str__(self, level=0):
        ret = "  " * level + repr(self.sp) + " " + Tree._show(self.closed) + "\n"
        for kid in self.kids:
            ret += kid.__str__(level+1)
        return ret

    @staticmethod
    def _show(b):
        return "{X}" if b else "{O}"



def isSat(tree):
    return not tree.closed

def allAtoms(sp):
    return len(list(filter(lambda x : not x.isAtom(), sp))) == 0

def complement(p):
    return p.p if getName(p) == "Not" else Not(p)

def existsComplements(sp):
    if len(sp) <= 1:
        return False

    m = findMin(sp)
    r = copy(sp)
    r.remove(m)
    return True if complement(m) in r else existsComplements(r)

def describeSet(sp):
    if existsComplements(sp):
        return SetStatus.Closed
    if not allAtoms(sp):
        return SetStatus.NonAtomic
    else:
        return SetStatus.Open


"""
w0: [p1, ..., pn]
- exhaust type-1 rules
- for each diamond(p) -> create new world where p is true
- box(q) will be copied in all newly created worlds
- K: p -> box(p), so all literals are also copied in all newly created worlds

in world w, exhaust all type1 rules

in world w, for each diamond(phi) in sp: --type2
    construct new world wi
    put phi in wi
    put all box(psi), p from sp in wi --type3
    link wi with w

    recursively on spi in wi
"""

def tableaux(phi):
    return build([phi])

def build(sp):
    desc = describeSet(sp)

    if desc == SetStatus.Open:
        return Tree(sp, False, [])

    if desc == SetStatus.Closed:
        return Tree(sp, True, [])

    if desc == SetStatus.NonAtomic:
        nonAtoms, _ = partition(lambda p : not p.isAtom(), sp)
        m = findMin(nonAtoms)
        sp2 = copy(sp)
        sp2.remove(m)
        rest = list(map(lambda x : build(sp2 + x), branchOn(m)))
        closed = reduce(lambda x, y: x and y, map(lambda n : n.closed, rest), True)
        return Tree(sp, closed, rest)

def alpha(sp):
    return [sp]

def beta(sp):
    return list(map(lambda x : [x], sp))

def branchOn(phi):
    if getName(phi) == 'And':
        return alpha([phi.p1, phi.p2])
    if getName(phi) == 'Not' and getName(phi.p) == 'Or':
        return alpha([Not(phi.p.p1), Not(phi.p.p2)])
    if getName(phi) == 'Not' and getName(phi.p) == 'Imp':
        return alpha([phi.p.p1, Not(phi.p.p2)])
    if getName(phi) == 'Iff':
        return alpha([Imp(phi.p1, phi.p2), Imp(phi.p2, phi.p1)])
    if getName(phi) == 'Not' and getName(phi.p) == 'Not':
        return alpha([phi.p.p])

    if getName(phi) == 'Or':
        return beta([phi.p1, phi.p2])
    if getName(phi) == 'Imp':
        return beta([Not(phi.p1), phi.p2])
    if getName(phi) == 'Not' and getName(phi.p) == 'And':
        return beta([Not(phi.p.p1), Not(phi.p.p2)])
    if getName(phi) == 'Not' and getName(phi.p) == 'Iff':
        return beta([Not(Imp(phi.p.p1, phi.p.p2)), Not(Imp(phi.p.p2, phi.p.p1))])
